vllm/ex/ex.py

```diff
@@ -0,0 +1,245 @@
+            return False
+
+        if node.op == 'call_method':
+            class_type = call_method_class(node)
+            return any([issubclass(class_type, c) for c in SUPPORTED[fn_name]])
+        else:
+            return True
+    else:
+        return False
+
+
+"""
+Partition 'gm' into submodules based on the 'is_node_supported' callback.
+Modules containing "supported" nodes will be optimized by the backend.
+"""
+def partition_graph(
+    gm: torch.fx.GraphModule,
+    example_inputs: List[torch.Tensor]
+) -> Tuple[torch.fx.GraphModule, List[Partition]]:
+    support = create_op_support(is_node_supported)
+
+    p = CapabilityBasedPartitioner(
+        gm,
+        support,
+        allows_single_node_partition=False,
+        non_compute_ops=None,
+        allowed_single_node_partition_ops=None
+    )
+    parts = p.propose_partitions()
+    part_gm = p.fuse_partitions(parts)
+
+    return part_gm, parts
+
+
+###############################################################################
+#
+# Inliner
+#
+###############################################################################
+
+"""
+Inline all submodules in 'mod' by running the tracer on them.
+TBD
+"""
+def inline_submodules(mod: torch.fx.GraphModule) -> torch.fx.GraphModule:
+    return mod
+
+
+###############################################################################
+#
+# Backend
+#
+###############################################################################
+
+"""
+Run optimizer on the given module.
+"""
+def optimize(
+    cc: CodeCache,
+    fgen: FusedOpGenerator,
+    mod: torch.fx.GraphModule,
+    example_inputs: List[torch.Tensor],
+) -> torch.fx.GraphModule:
+    mod = rewrite_quantized_gemms(mod, example_inputs)
+    mod = pointwise_fusion(cc, fgen, mod, example_inputs)
+    # TODO: should we re-trace here to inline?  or will inductor handle it?
+    # mod = inline_submodules(mod)
+    return mod
+
+
+"""
+Compile a module with the given backend.
+"""
+def backend_compile(
+    gm: torch.fx.GraphModule,
+    example_inputs: List[torch.Tensor],
+    backend: str ='inductor'
+) -> Callable:
+    try:
+        backend = lookup_backend(backend)
+        logger.debug(f"attempting {backend} on {gm.name}")
+        backend_compiled = backend(gm, example_inputs)
+        if backend_compiled is not None:
+            logger.debug(f"{backend} compiled {gm.name}.")
+            return backend_compiled
+    except Exception as ex:
+        logger.warning(f"backend_compile failed: {ex}")
+        logger.warning(f"Trace: {traceback.format_tb(ex.__traceback__)}")
+        pass
+
+    return gm.forward
+
+
+def node_in_module(n: torch.fx.Node, m: torch.fx.GraphModule) -> bool:
+    # Note: this doesn't work: return n.graph.owning_module == m
+    # Names should be unique, so this is ok
+    return n.name in [nn.name for nn in m.graph.nodes]
+
+
+def module_in_partitions(parts: List[Partition], m: torch.fx.GraphModule) -> bool:
+    for p in parts:
```

Comment:
`# Names should be unique so if the first node in the partition exists in m that's enough to confirm that m is in that partition`?